File: cpp/codeql-dataflow-sql-injection/add-user.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <sqlite3.h>
#include <time.h>

void write_log(const char* fmt, ...) {
    time_t t;
    char tstr[26];
    va_list args;

    va_start(args, fmt);
    t = time(NULL);
    ctime_r(&t, tstr);
    tstr[24] = 0; /* no \n */
    fprintf(stderr, "[%s] ", tstr);
    vfprintf(stderr, fmt, args);
    va_end(args);
    fflush(stderr);
}

void abort_on_error(int rc, sqlite3 *db) {
    if( rc ) {
        fprintf(stderr, "Can't open database: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        fflush(stderr);
        abort();
    }
}

void abort_on_exec_error(int rc, sqlite3 *db, char* zErrMsg) {
    if( rc!=SQLITE_OK ){
        fprintf(stderr, "SQL error: %s\n", zErrMsg);
        sqlite3_free(zErrMsg);
        sqlite3_close(db);
        fflush(stderr);
        abort();
    }
}
    
char* get_user_info() {
#define BUFSIZE 1024
    char* buf = (char*) malloc(BUFSIZE * sizeof(char));
    if(buf==NULL) abort();
    int count;
    // Disable buffering to avoid need for fflush
    // after printf().
    setbuf( stdout, NULL );
    printf("*** Welcome to sql injection ***\n");
    printf("Please enter name: ");
    count = read(STDIN_FILENO, buf, BUFSIZE - 1);
    if (count <= 0) abort();
    // ensure the buffer is zero-terminated
    buf[count] = '\0';
    /* strip trailing whitespace */
    while (count && isspace(buf[count-1])) {
        buf[count-1] = 0; --count;
    }
    return buf;
}

int get_new_id() {
    int id = getpid();
    return id;
}

void write_info(int id, char* info) {
    sqlite3 *db;
    int rc;
    int bufsize = 1024;
    char *zErrMsg = 0;
    char query[bufsize];
    
    /* open db */
    rc = sqlite3_open("users.sqlite", &db);
    abort_on_error(rc, db);

    /* Format query */
    snprintf(query, bufsize, "INSERT INTO users VALUES (%d, '%s')", id, info);
    write_log("query: %s\n", query);

    /* Write info */
    rc = sqlite3_exec(db, query, NULL, 0, &zErrMsg);
    abort_on_exec_error(rc, db, zErrMsg);

    sqlite3_close(db);
}

int main(int argc, char* argv[]) {
    char* info;
    int id;
    info = get_user_info();
    id = get_new_id();
    write_info(id, info);
     free(info);
    /*
     * show_info(id);
     */
}

File: python/codeql-dataflow-sql-injection/add-user.py
#!/usr/bin/env python3

import argparse
import logging
import os
import sqlite3

class EmptyNameError(Exception):
    pass

logging.basicConfig(force=True, encoding='utf-8', level=logging.INFO,
                    format='[%(asctime)s] %(message)s', datefmt='%Y-%m-%d %H:%M:%S')

def get_user_info():
    print("*** Welcome to sql injection ***")
    print("Please enter name: ", end='', flush=True)
    info = input().strip()
    if info == "": raise EmptyNameError("get_user_info: expect non-empty name")
    return info

def get_new_id():
    id = os.getpid()
    return id

def write_info(id: int, info: str) -> None:
    # Open db
    conn = sqlite3.connect("users.sqlite")

    # Format query 
    query = "INSERT INTO users VALUES (%d, '%s')" % (id, info)

    # Write data
    conn.executescript(query)     # Unsafe, used for illustration
    logging.info("query: %s\n", query)

    # Finish up
    conn.commit()
    conn.close()

def add_user():
    # Command line interface
    parser = argparse.ArgumentParser(
        description="""A server-side interface to add users to the database.  
A single user name is read from stdin, actions are logged to stderr""",
        formatter_class=argparse.RawDescriptionHelpFormatter
    )
    parser.parse_args()

    # Read and process input
    logging.info("Running add-user")
    info = get_user_info()
    id = get_new_id()
    write_info(id, info)

if __name__ == "__main__":
    add_user()
    
